### PR TITLE
start algebraic extensions

### PR DESCRIPTION
**Pull request by tthsqe12**
There was one remaining important coefficient ring provided natively by singular: algebraic extensions of Q and Fp. Here is a taste of how it works.
Needs https://github.com/oscar-system/libsingular-julia/pull/11

```
julia> F,(a,)=FunctionField(QQ, ["a"])
(Function Field over Rational Field with transcendence basis n_transExt[(a)], n_transExt[(a)])

julia> F,a=AlgebraicExtensionField(F,a^2+1)
(Algebraic Extension of Rational Field with defining equation (a^2+1), (a))

julia> (3*a+1)//(2*a+1)
(1/5*a+7/5)

julia> R,(x,y,z) = PolynomialRing(F, ["x", "y", "z"])
(Singular Polynomial Ring (0,a),(x,y,z),(dp(3),C), spoly{n_algExt}[x, y, z])

julia> (a*x+y+z)^3
-a*x^3 - 3*x^2*y + 3*a*x*y^2 + y^3 - 3*x^2*z + 6*a*x*y*z + 3*y^2*z + 3*a*x*z^2 + 3*y*z^2 + z^3


julia> F,(a,)=FunctionField(Fp(7), ["a"])
(Function Field over Finite Field of Characteristic 7 with transcendence basis n_transExt[(a)], n_transExt[(a)])

julia> F,a=AlgebraicExtensionField(F,a^2+1)
(Algebraic Extension of Finite Field of Characteristic 7 with defining equation (a^2+1), (a))

julia> (3*a+1)//(2*a+1)
(3*a)

julia> R,(x,y,z) = PolynomialRing(F, ["x", "y", "z"])
(Singular Polynomial Ring (7,a),(x,y,z),(dp(3),C), spoly{n_algExt}[x, y, z])

julia> (a*x+y+z)^3
-a*x^3 - 3*x^2*y + 3*a*x*y^2 + y^3 - 3*x^2*z - a*x*y*z + 3*y^2*z + 3*a*x*z^2 + 3*y*z^2 + z^3
```

### CODE REVIEW

File: src/number/n_algExt.jl [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+*(x::Integer, y::n_algExt) = parent(y)(x)*y
+
+###############################################################################
+#
+#   Comparison
+#
+###############################################################################
+
+function ==(x::n_algExt, y::n_algExt)
+    return libSingular.n_Equal(x.ptr, y.ptr, parent(x).ptr)
```
Comment (thofma) [inline src/number/n_algExt.jl:180]: Does this need a prent check first?

**Comment by thofma** [comment]
Test seem to be passing without https://github.com/oscar-system/libsingular-julia/pull/11?

**Comment by tthsqe12** [comment]
That is because there are no tests yet

File: src/number/n_algExt.jl (hunk 1 above)
Comment (tthsqe12) [reply src/number/n_algExt.jl:180]: None of the other number types check the parents. I will open an issue if this is a problem.

**Comment by thofma** [comment]
I wondered about this too. But I think this is actually a residue field, the residue field of the function field `F = Q(t)` at the prime ideal defined by `a^2+1`.

**Comment by tthsqe12** [comment]
It is not really a residue field of the function field, but that is the way to construct it in singular.
consider AA:
```
julia> Zxy, (x, y) = ZZ["x", "y"]
(Multivariate Polynomial Ring in x, y over Integers, AbstractAlgebra.Generic.MPoly{BigInt}[x, y])

julia> Qxy = FractionField(Zxy)
Fraction field of Multivariate Polynomial Ring in x, y over Integers

julia> x = Qxy(x); y = Qxy(y);

julia> x
x

julia> typeof(ans)
AbstractAlgebra.Generic.Frac{AbstractAlgebra.Generic.MPoly{BigInt}}

julia> F = ResidueField(Qxy, x^2-y)
Residue field of Fraction field of Multivariate Polynomial Ring in x, y over Integers modulo 1

julia> F(x)
ERROR: MethodError: no method matching mod(::AbstractAlgebra.Generic.Frac{AbstractAlgebra.Generic.MPoly{BigInt}}, ::AbstractAlgebra.Generic.Frac{AbstractAlgebra.Generic.MPoly{BigInt}})
```

**Comment by thofma** [comment]
I know that it is not a quotient ring of the field itself, but in the language of function fields it is the residue field at the place defined by the irreducible polynomial.

**Comment by mohamed-barakat** [comment]
It is the field of fractions of a residue class ring Z[x,y] / an irreducible ideal. Why shouldn't it be constructed that way and why shouldn't the output comply with the correct mathematical language?

**Comment by fieker** [comment]
On Mon, Feb 01, 2021 at 07:12:21AM -0800, Mohamed Barakat wrote:
> It is the field of fractions of a residue class ring Z[x,y] / an irreducible ideal. Why shouldn't it be constructed that way and why shouldn't the output comply with the correct mathematical language?
> 

Two thoughts:
 - Oscar: you're right, and here matheamtics should reign supreme
 - Singular.jl: I don't know if we can affort to "fix" all the
   idiosyncrasis of the underlying Singular, thus am not sure if
   Singular.jl should deviate too much from being an interface to
   Singular itself
>
Otherwise: you're right

**Comment by mohamed-barakat** [comment]
Ah, this is a Singular, not an OSCAR PR, sorry. Yes, I agree, as long as the end-user is never confronted with it.

File: src/number/NumberTypes.jl [hunk 2]
```diff
@@ -317,6 +322,71 @@ n_transExt(c::N_FField, n::Nemo.fmpz) = n_transExt(c, libSingular.number_ptr(n,
...
 
 
+###############################################################################
+#
+#   SingularAlgebraicExtensionField/n_algExt
+#
+###############################################################################
+
+const N_AlgExtFieldID = Dict{Tuple{N_FField, n_transExt}, Field}()
+const N_AlgExtFieldIDptr = Dict{libSingular.coeffs_ptr, Field}()
```
Comment (tthsqe12) [inline src/number/NumberTypes.jl:332]: @thofma could you take a look a my caching here?

Comment (thofma) [reply src/number/NumberTypes.jl:332]: Do we need to always cache it?

Comment (tthsqe12) [reply src/number/NumberTypes.jl:332]: I have no idea. I think they should be cached be default, and the ones that come from singular should be cached. But, none of the other constructors have a cached argument. I should add the cached = keyword argument?

Comment (thofma) [reply src/number/NumberTypes.jl:332]: Usually the user facing ones have the option to disable the caching. I guess some constructors here must have the caching to make it work with Singular? I don't know enough about the details.

Comment (tthsqe12) [reply src/number/NumberTypes.jl:337]: I added cached for the ones that face the user

**Comment by tthsqe12** [comment]
Well, this took longer than expected, but I think it is good now.

File: src/number/n_algExt.jl [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+function characteristic(R::N_AlgExtField)
+   return ZZ(libSingular.n_GetChar(R.ptr))
+end
+
+function deepcopy_internal(a::n_algExt, dict::IdDict)
+   return parent(a)(libSingular.n_Copy(a.ptr, parent(a).ptr))
+end
+
+function hash(a::n_algExt, h::UInt)
+   error("hash not implemented")
```
Comment (thofma) [inline src/number/n_algExt.jl:42]: You sure about this one?

File: src/number/n_algExt.jl [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+#
+#   AlgebraicExtensionField constructor
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    AlgebraicExtensionField(F::Singular.N_FField, a::n_transExt)
+
+Given a function field F = R(x) in one variable and a defining polynomial a
+in R[x], return a tuple consisting of the new field R[x]/(a) and x.
```
Comment (thofma) [inline src/number/n_algExt.jl:321]:
```suggestion
in R[x], return a tuple consisting of the new field R[x]/(a) and the class of x.
```

File: src/number/n_algExt.jl (hunk 3 above)
Comment (tthsqe12) [reply src/number/n_algExt.jl]: proper implementation needs more help from the binaries. maybe next time around.

Comment (thofma) [reply src/number/n_algExt.jl]: Making it not crash is a good start.

**Pull request merged**